add static interfaces in the form of signature types

zerbina opened the pull request:
Signature types are meant to provide static interfaces, which would be a first step towards type-checked generic routines and (maybe) eventually removing `.borrow`.

**The name `signature` is a working title.** It's not yet clear whether the feature will take over `concept` or stay as its own entity, and thus the `concept` syntax is simply reused.

**The current specification is not meant to represent the final form.** It's a version of the feature as discussed so far that's narrowed-down as much as possible in order to make a robust implementation feasible/possible with the current compiler. Further restrictions due to issues discovered during implementation are likely.

The feature was inspired by the `signature` feature of Standard ML.

---

## To-Do
- [x] ~add a signature pragma~
- [x] add dedicated `TType` kinds for signature types and their applied form 
- [ ] implement signature body analysis (partially done)
- [x] implement argument-dependent lookup (for signature types)
- [x] implement signature application (will be based on the existing borrow mechanism, most likely)
- [ ] add tests
- [ ] write a proper commit message

zerbina committed:
manual: add a first draft of the signature specification

beef331 commented on doc/manual.rst:
```diff
@@ -4166,6 +4183,85 @@ For the above `openArray` and `var` rules, whether the expressions refers to
+and `macro`:idx: are not.
+
+For every routine declaration in the body, at least one parameter must be of
+type `placholder`, `var placeholder`, or `sink placeholder`. Declarations must
```
`placholder` -> `placeholder`

beef331 commented on doc/manual.rst:
```diff
@@ -4166,6 +4183,85 @@ For the above `openArray` and `var` rules, whether the expressions refers to
+or more candidates that satisfy the requirements listed above, application
+fails. If sucessful, the resulting application is said to *bind* to `y`.
+
+Two applied signatures are equal are equal iff both are applications of the
```
`iff` -> `if`

zerbina replied:
The spelling is deliberate. "iff" is a short form of "if and only if" ([source](https://en.wikipedia.org/wiki/If_and_only_if)).

beef331 replied:
Perhaps it's not shorter if you have to provide this message along with all reading material :stuck_out_tongue:

zerbina replied:
Well, `iff` is already used in other places of the manual.

Nonetheless, you're right in that the usage does make the text less accessible; I'll expand it.

beef331 commented on doc/manual.rst:
```diff
@@ -4166,6 +4183,85 @@ For the above `openArray` and `var` rules, whether the expressions refers to
+On application of `S` to `T`, for every routine declaration `x` of `S`, a
+routine `y` must be *visible* in the current context such that:
+* `x` and `y` have the same name
+* after all `placholder`s in the parameter list of `x` are replaced with `T`,
```
`placholder` -> `placeholder`

zerbina committed:
Merge branch 'devel' into signature-types

zerbina committed:
semtypinst: handle signature types (preliminary)

zerbina committed:
typesrenderer: implement preliminary handling for signature types

zerbina committed:
vmdeps: implement preliminary rendering of signature types

zerbina committed:
add `tySignature` and `tySignatureInst`

To represent signature types and applied signature types, respectively.

zerbina committed:
sem: implement argument-dependent lookup

zerbina committed:
liftdestructors: handle signature types

In the context of hooks, applied signature types are ignored and treated
as if they were their underlying type.

zerbina committed:
manual: make feature adjustments & improve some wording

* there's a dedicated `signature` type constructor now. This is better
  than the concept + pragma approach, but still not ideal
* `method`s are no longer allowed in signatures. I couldn't think of a
  good reason for allowing them ("why not" doesn't count)
* call operators are now allowed. This rule might need to be narrowed
  down again

zerbina committed:
typeallowed: handle signature types

Applied signature types are concrete types and may thus be used
anywhere concrete types may be used; raw signature types are meta types.

zerbina committed:
semtypes: implement first version of `signature` operator

zerbina committed:
jsgen: handle signature types

zerbina committed:
types: handle signature types w.r.t. type traits

zerbina committed:
implement type relations for signature types

zerbina committed:
semtypes: introduce type variable for `tySignature` parameters

zerbina committed:
mirgen: handle signature types

They're handled much like concepts, in that they're treated as their
underlying type.

zerbina committed:
transf: support borrowed iterators

zerbina committed:
tests: add a first batch of specification tests

They're not good, but neither are most of the existing specification
tests...

zerbina committed:
sigmatch: use proper symbol lookup for signatures

Picking a symbol to borrow via regular overload resolution is wrong
because:
* it cannot enforce specific effects for the overload (raises, tags,
  and noSideEffect)
* it admits conversions
* it effectively ignores `sink` and other type modifiers

all of which are things that are not desired when picking a routine to
borrow from.

zerbina committed:
semtypes: progress towards working effects on signature requirements

* handle pragma lists
* set up the types w.r.t. effects
* set the correct default calling convention for iterators
